lib/matplotlib/axes/_axes.py

```diff
@@ -2309,14 +2330,24 @@ def bar(self, x, height, width=0.8, bottom=None, *, align="center",
         # subtracted uniformly
         if self.xaxis is not None:
             x0 = x
-            x = np.asarray(self.convert_xunits(x))
+            # self.convert_xunits doesn't convert Decimal input, use
+            # _decimal_convert to force conversion
+            x = self._decimal_convert(self.convert_xunits(x))
```

Comment:
I'm not sure about the design here. On the one hand, needing a converter around the converter feels wrong. On the other hand, do we want to change the return type of `convert_xunits()`?

Reply:
I agree that the design seems(?) problematic (is this converter going to end up being needed *everywhere*?).

Reply:
Thanks for advice @timhoffm @anntzer . But I am not sure whether I should update convert_units(self, x) method in Axis directly.

convert_xunits mainly consumes convert_units(self, x) in Axis class, whose first line is:
```
# If x is already a number, doesn't need converting
if munits.ConversionInterface.is_numlike(x):
      return x
```
Since Decimal is actually a Number, this method doesn't change the input x if x is a list of Decimal. 

One possible solution is like:
```
# If x is already a number, doesn't need converting
if munits.ConversionInterface.is_numlike(x):
      return np.asarray(x, dtype=np.float)
```
But I am afraid this way may cause side effect when input x is a list of integer or other number type. 

If we check whether x contains Decimal to decide whether need to convert x, then I feel the is_numlike(x) function in ConversionInterface doesn't quite make sense here.

W

Reply:
Besides, I am curious about the following code in is_numlike(x):
```
if iterable(x):
    for thisx in x:
        return isinstance(thisx, Number)
```
This only check the first element in a list. Is this what it supposed to do when input x is a list?

Reply:
Perhaps is_numlike shouldbe changed to `isinstance(x, Number) and not isinstance(x, Decimal)` then (dunno).
The unit system assumes that lists are homogeneous.  Yes, it's a pretty awkward assumption.

Reply:
In my newest push, I added checking and converting for Decimal in convert_units() of Axis, since I think is_numlike make sense to stand there. Since unit system assumes that lists are homogeneous, I chose to keep this assumption in my checking. Would you @anntzer mind having a check and see whether it make sense?

Reply:
I haven't spent a lot of time to think about it, but perhaps the correct approach may be to remove the `if is_numlike(x): return x` in `convert_units()` and instead always go through the unit conversion system, and then add a converter for `Number` (which will thus be used both for e.g. floats and for Decimal) which casts the input to float.

Reply:
If we add a converter for Number, will the precision of integer input be affected, especially for big number?

Reply:
Also I am not sure where would be the appropriate place to add the code for Number converter?What should be the return value of axisinfo(unit, axis) and default_units(x, axis) for this Number converter?

Reply:
I doubt the precision matters as everything ultimately gets converted to float at some point.

The default converter would be implemented in units.py but I certainly haven't thought about the specifics...

Reply:
> I haven't spent a lot of time to think about it, but perhaps the correct approach may be to remove the `if is_numlike(x): return x` in `convert_units()` and instead always go through the unit conversion system, and then add a converter for `Number` (which will thus be used both for e.g. floats and for Decimal) which casts the input to float.

On first thought, it sounds reasonable to not special-case numbers but just handle them as a united quantity. However, I'm not sure if that would work:
~~~
    def convert_units(self, x):
        # If x is already a number, doesn't need converting
        if munits.ConversionInterface.is_numlike(x):
            return x

        if self.converter is None:
            self.converter = munits.registry.get_converter(x)
~~~

If `self.converter` is not `None`, you wouldn't get the proposed `NumberFormatter`; and therefore all converters would be required to pass numbers through`.

Reply:
@sasoripathos Sorry that we're quite slow and vague with responses here. You've chosen a tricky issue; and as you can see from the discussion, we do not have a clear idea what the appropriate fix should be.

Reply:
@timhoffm Thanks for reply! That's also a place I am confused about. I still haven't figure out what should be the return value of axisinfo(unit, axis) and default_units(x, axis) if i implemented the proposed NumberFormatter. Please pardon my beginner knowledge about matplotlib. I am still looking into the proposed approach and will give a try next week. I also think my current implementation is not that elegent since it (kind of) "breaks" the Number type, but i think it may be (hopefully) a working walk around.

Reply:
@timhoffm @anntzer @jklymak Hi! I just pushed a DecimalConverter I implemented. Since originally no convertion will be done if x is a number-like input, I just make axisinfo(unit, axis) return empty AxisInfo and default_units(x, axis) return None. Besides, I use ConversionInterface.is_numlike(x) to avoid requiring all existing converters pass Numbers. 

Could you please have a review and give me some feedback? Thanks very much!